src/coreComponents/physicsSolvers/solidMechanics/kernels/StrainHelper.hpp

```diff
@@ -214,6 +214,182 @@ class AverageStrainOverQuadraturePointsKernelFactory
 
 
 
+
+
+
+/**
+ * @class AverageStressOverQuadraturePoints
+ * @tparam SUBREGION_TYPE the subRegion type
+ * @tparam FE_TYPE the finite element type
+ */
+template< typename SUBREGION_TYPE,
+          typename FE_TYPE >
+class AverageStressOverQuadraturePoints :
+  public AverageOverQuadraturePointsBase< SUBREGION_TYPE,
+                                          FE_TYPE >
+{
+public:
+
+  /// Alias for the base class;
+  using Base = AverageOverQuadraturePointsBase< SUBREGION_TYPE,
+                                                FE_TYPE >;
+
+  using Base::m_elementVolume;
+  using Base::m_elemsToNodes;
+  using Base::m_finiteElementSpace;
+
+  /**
+   * @brief Constructor for the class
+   * @param nodeManager the node manager
+   * @param edgeManager the edge manager
+   * @param faceManager the face manager
+   * @param elementSubRegion the element subRegion
+   * @param finiteElementSpace the finite element space
+   * @param stress the stress solution field
+   * @param avgStress the stress averaged over quadrature points
+   */
+  AverageStressOverQuadraturePoints( NodeManager & nodeManager,
+                                     EdgeManager const & edgeManager,
+                                     FaceManager const & faceManager,
+                                     SUBREGION_TYPE const & elementSubRegion,
+                                     FE_TYPE const & finiteElementSpace,
+                                     arrayView3d< real64 const, solid::STRESS_USD > const stress,
+                                     fields::solidMechanics::arrayView2dLayoutAvgStress const avgStress ):
+    Base( nodeManager,
+          edgeManager,
+          faceManager,
+          elementSubRegion,
+          finiteElementSpace ),
+    m_stress( stress ),
+    m_avgStress( avgStress )
+  {}
+
+  /**
+   * @copydoc finiteElement::KernelBase::StackVariables
+   */
+  struct StackVariables : Base::StackVariables
+  {  };
+
+  /**
+   * @brief Performs the setup phase for the kernel.
+   * @param k The element index.
+   * @param stack The StackVariable object that hold the stack variables.
+   */
+  GEOS_HOST_DEVICE
+  void setup( localIndex const k,
+              StackVariables & stack ) const
+  {
+    Base::setup( k, stack );
+
+    for( int icomp = 0; icomp < 6; ++icomp )
+    {
+      m_avgStress[k][icomp] = 0.0;
+    }
+  }
+
+  /**
+   * @brief Increment the average property with the contribution of the property at this quadrature point
+   * @param k The element index
+   * @param q The quadrature point index
+   * @param stack The StackVariables object that hold the stack variables.
+   */
+  GEOS_HOST_DEVICE
+  void quadraturePointKernel( localIndex const k,
```

Comment:
This file contains two functions `quadraturePointKernel` once for the strain and one for the stress. In my opinion only one function should exist to perform the averaging of a second order symmetric stress tensor -- and probably of any scalar (rank 0), vector (rank 1), ... . The class [`AverageOverQuadraturePointsBase`](https://github.com/GEOS-DEV/GEOS/blob/93b3a72a965fad4cc349674286bafb08761dac86/src/coreComponents/mesh/utilities/AverageOverQuadraturePointsKernel.hpp#L38)  should be responsible for this task, and the quantity to be averaged should either provided directly or with a lambda functions that evaluates such property.

Reply:
we could generalize it and come up with a single kernel. We could try to make it accept a list of views instead of a specific one. It would also mean that we do all averaging in a single kernel launch which it is indeed better. We could use something similar to the `elementViewAccessors` and create an `array1d` of all scalars, one for all vectors and one for all tensors to be averaged and have the `averagingKernel` only ask for the keys of the fields that need to be averaged. 
The strain probably remains a bit of a special case though and i don't know that we do any averaging for other quantities.